File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 6/11/2026, 11:38:31 AM
class Solution:
    def isValid(self, s: str) -> bool:
        store = {"{":"}","[":"]","(":")"}

        stack = []
        for i in range(len(s)):
            if (s[i] in store):
                stack.append(s[i])
            else:
                if(stack == []):
                    return False
                else:
                    if(store[stack.pop()] != s[i]):
                        return False
        if(stack != []):
            return False
        return True

File: 20-ValidParentheses/20-ValidParentheses.cpp
// Last updated: 6/11/2026, 11:38:25 AM
class Solution {
public:
    unordered_map<char, char> ump;
    bool isValid(string s) {
        ump['{'] = '}';
        ump['('] = ')';
        ump['['] = ']';

        stack<char> st;

        for(char ch: s){
            if(ump.find(ch) != ump.end()){
                st.push(ch);
            }else{
                if(!st.empty() && ump[st.top()] == ch){
                    st.pop();
                    continue;
                }else{
                    return 0;
                }
            }
        }

        return (st.empty() == 1);

    }
};